syntax = "proto3";

message CmdGetStatus {
    string cmd = 1;
    int32 mode = 2;
}

message RespGetStatus {
    string resp = 1;
    int32 mode = 2;
}

enum EncryptionMode {
    Type_Open = 0;
    Type_WEP = 1;
    Type_WPA_PSK = 2;
    Type_WPA2_PSK = 3;
    Type_WPA_WPA2_PSK = 4;
    Type_WPA2_ENTERPRISE = 5;
    Type_WPA3_PSK = 6;
    Type_WPA2_WPA3_PSK = 7;
}

message CmdConfig {
    string ssid = 1;
    string pwd = 2;
    EncryptionMode ecn = 3;
    int32 chnl = 4;
    int32 max_conn = 5;
    bool ssid_hidden = 6;
    string bssid = 7;
    int32 rssi = 8;
    string status = 9;
    int32 bw = 10;
    bool is_wpa3_supported = 11;
}

message RespConfig {
    string ssid = 1;
    string pwd = 2;
    EncryptionMode ecn = 3;
    int32 chnl = 4;
    int32 max_conn = 5;
    bool ssid_hidden = 6;
    string bssid = 7;
    int32 rssi = 8;
    string status = 9;
    int32 bw = 10;
    bool is_wpa3_supported = 11;
}

message ScanResult {
    bytes ssid = 1;
    uint32 chnl = 2;
    int32 rssi = 3;
    bytes bssid = 4;
    EncryptionMode ecn = 5;
}

message CmdScanResult {
}

message RespScanResult {
    uint32 count = 1;
    repeated ScanResult entries = 2;
}

message ConnectedSTAList {
	bytes mac = 1;
	int32 rssi = 2;
}

message CmdConnectedSTA {
}

message RespConnectedSTA {
	uint32 num = 1;
	repeated ConnectedSTAList stations = 2;
}

enum EspHostedConfigMsgType {
    TypeCmdGetMACAddress = 0;
    TypeRespGetMACAddress = 1;
    TypeCmdGetWiFiMode = 2;
    TypeRespGetWiFiMode = 3;
    TypeCmdSetWiFiMode = 4;
    TypeRespSetWiFiMode = 5;
    TypeCmdGetAPConfig = 6;
    TypeRespGetAPConfig = 7;
    TypeCmdSetAPConfig = 8;
    TypeRespSetAPConfig = 9;
    TypeCmdGetSoftAPConfig = 10;
    TypeRespGetSoftAPConfig = 11;
    TypeCmdSetSoftAPConfig = 12;
    TypeRespSetSoftAPConfig = 13;
    TypeCmdDisconnectAP = 14;
    TypeRespDisconnectAP = 15;
    TypeCmdGetAPScanList = 16;
    TypeRespGetAPScanList = 17;
    TypeCmdGetConnectedSTAList = 18;
    TypeRespGetConnectedSTAList = 19;
}

message EspHostedConfigPayload {
    EspHostedConfigMsgType msg = 1;
    oneof payload {
        CmdGetStatus cmd_get_mac_address = 10;
        RespGetStatus resp_get_mac_address = 11;
        CmdGetStatus cmd_get_wifi_mode = 12;
        RespGetStatus resp_get_wifi_mode = 13;
        CmdGetStatus cmd_set_wifi_mode = 14;
        RespGetStatus resp_set_wifi_mode = 15;
        CmdConfig cmd_get_ap_config = 16;
        RespConfig resp_get_ap_config = 17;
        CmdConfig cmd_set_ap_config = 18;
        RespConfig resp_set_ap_config = 19;
        CmdConfig cmd_get_softap_config = 20;
        RespConfig resp_get_softap_config = 21;
        CmdConfig cmd_set_softap_config = 22;
        RespConfig resp_set_softap_config = 23;
        CmdGetStatus cmd_disconnect_ap = 24;
        RespGetStatus resp_disconnect_ap = 25;
        CmdScanResult cmd_scan_ap_list = 26;
        RespScanResult resp_scan_ap_list = 27;
        CmdConnectedSTA cmd_connected_stas_list = 28;
        RespConnectedSTA resp_connected_stas_list = 29;
    }
}
